joomla/joomla-cms
# [4.0] Support External Adapters For Media Form Field

Pull Request for Issue #26750.

### Summary of Changes
This is my quick attempt to solve the issue with external adapters discussed in https://github.com/joomla/joomla-cms/issues/26750. The change includes:

- For local adapter, it will just work as how it was before 
- For external adapters:
+ Store full url of the image into value of the field so that we do not have to ask the adapter to get the image url each time the image is being displayed
+ The adapter + the internal path of the image is stored as part of value, too. It is added to the end of the image URL (# + Adapter Name + Internal Path). I decided to use # character instead of & character because Dropbox does not allow adding query string to image url.
- Modify HTMLHelper to remove the adater name + Internal Path from the image before it is displayed


### Testing Instructions
1. Download update package for this PR at https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/33724/downloads/44325/Joomla_4.0.0-beta8-dev+pr.33724-Development-Update_Package.zip . Go to System -> Update -> Joomla, upload and update this package.
2. General testing is test and make sure you can select image for your articles and these selected images will still be displayed properly when you view the article from frontend of your site.

###  Test local adapter:
- Create a article, select an image for that article, make sure you can still select the image. Save the article .
- Access to the article from frontend, make sure the image is being displayed
- Edit that article again. Click on the Select button next to the image and try to select new image. Please make sure that you don't see any error while clicking on that Select button. Also, click on that Select button should show you the folder which the image belong to (basically, the system remember the path of the image, and you do not have to navigate from root to find a new image.

### Test External Adapter
It will be good to test media manager with an external adapter like dropbox and make sure it is working OK. You can"

- Download Dropbox Adapter plugin from this link https://github.com/joomla/joomla-cms/files/6526439/plug_filesystem_dpdropbox.zip . Go to System -> Extensions -> Install to install that plugin
- Go to System -> Manage -> Plugins,  find that plugin, click on it to edit and enter necessary parameters. You can find the testing parameters from my Dropbox account below:

Client ID : 2p8aoyqpeub1a39
Client Secret: 2p8aoyqpeub1a39
Access Token: jeU8dH-JufAAAAAAAAAAAex3adAkjxBV5e7BqzrOi6T60ZRzwhVQUNJDOnmitGOa

![Plugins-DPDropbox-Joomla-4-Administration](https://user-images.githubusercontent.com/977664/119339715-b8c8b680-bcbb-11eb-860d-0a3dd342c7fa.png)

Then try to create article or edit article, select an image from dropbox adapter for that article, save the article. Then view that article from frontend, make sure the image is still being displayed.
